src/lib.rs

```diff
@@ -127,13 +128,16 @@ pub enum PingError {
     LibOpingError(String),
     /// A `std::ffi::NulError` that occurred while trying to convert a hostname string
     NulByteError(NulError),
+    /// An error with a custom error message
+    CustomError(String),
```

Comment:
I don't necessarily like adding a catch-all "custom" error: the type isn't meant to include other code's errors as well, just the errors that the library itself can produce.

I think that the above suggestions to add context to errors might give you what you need in your use case, but in general, the idiomatic way to add errors in a higher layer of the program is to wrap it in your own type. E.g., if your program had a need to return `OtherError(...)` from its `do_stuff()` function, you could define an enum:

    pub enum MyError {
        Ping(PingError),
        OtherError(...),
    }

You can then do something like `ping.send().map_err(|e| MyError::Ping(e))`, and this converts a `Result<(), PingError>` to a `Result<(), MyError>`, suitable for use in your `try!(...)` or whatnot.

The `error_chain` crate automates a lot of this, both the definition and some nice trait definitions that allow for auto-wrapping.